### PR TITLE
66 advanced deployment

### LINKED ISSUE #66: Advanced deployment
We need to add advanced deployment options (maybe in a disclosure):

- `sources` / `targets`
- filter and filter type
- restart

### PR DESCRIPTION
**Pull request by PaulFarault**
#### Which issue(s) this PR fixes

Fixes #66 

#### Agreements

- [X] I hereby declare this contribution to be licensed under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0).
- [X] I hereby agree to grant [TOSIT](https://www.tosit.io/) a copyright license to use my contributions.

### CODE REVIEW

File: src/components/TextArea.tsx [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+            'flex flex-wrap gap-1 border border-gray-300 rounded-md p-1 cursor-text',
+            isDisabled && 'bg-gray-200'
+          )}
+        >
+          {operations.map((v, i) => (
+            <Operation
+              key={v}
+              operation={v}
+              index={i}
+              {...{
```
Comment (gboutry) [inline src/components/TextArea.tsx:111]: Why is this syntax needed ? You are creating a new object from destructuring the old one, do you need a new ref ?

File: src/pages/deploy/index.tsx [hunk 2]
```diff
@@ -3,27 +3,115 @@ import DashboardLayout from 'src/app/dashboard/layout'
...
+          >
+            <div className="flex flex-col justify-center h-full items-center gap-2">
+              <div className="flex flex-col max-w-xs">
+                <OperationsList />
+                <InputField />
+              </div>
+              <Button onClick={() => deploy(operations)}>Deploy</Button>
+            </div>
+          </OperationsContextProvider> */}
+    </form>
```
Comment (gboutry) [inline src/pages/deploy/index.tsx:114]: Why is this the deploy button commented ?

File: src/pages/deploy/index.tsx [hunk 3]
```diff
@@ -3,27 +3,115 @@ import DashboardLayout from 'src/app/dashboard/layout'
...
+              placeholder="Filter"
+            />
+          </div>
+          <button
+            type="submit"
+            className="relative -ml-px inline-flex items-center space-x-2 rounded-r-md border border-gray-300 bg-gray-50 px-4 py-2 text-sm font-medium text-gray-700 hover:bg-gray-100"
+          >
+            <PlusIcon className="h-5 w-5 text-gray-400" aria-hidden="true" />
+            <span>Add</span>
+          </button>
```
Comment (gboutry) [inline src/pages/deploy/index.tsx:80]: The server supports different kinds of filter: Regex or Glob, we implement that here

File: src/components/TextArea.tsx (hunk 1 above)
Comment (PaulFarault) [reply src/components/TextArea.tsx:111]: First `{` is for the template, I'm destructuring only once ;)

**Review by gboutry — COMMENTED** [review]
Great work, additionnal remarks:
- Operation_name:
  * Pressing spacebar in the middle will confirm
  * We cannot remove the generated input field when its empty
  * Clicking outside leaves the field editable, therefore I'm not sure if my input is taken into account
- Filter:
  * I believe we should be able to at least validate a regex
- Restart:
  * Restart should be set to False by default

The validation toast could at least show the deployment id.

File: src/pages/deploy/index.tsx [hunk 4]
```diff
@@ -1,29 +1,97 @@
...
+      <FilterField
+        filterTypes={filterTypes}
+        filter={filter}
+        setFilter={setFilter}
+        filterType={filterType}
+        setFilterType={setFilterType}
+      />
+      <RestartField restart={restart} setRestart={setRestart} />
+      <div className="self-center">
+        <Button type="button" onClick={handleDeploy}>
```
[the remark this answers is not in the document]
Comment (PaulFarault) [reply src/pages/deploy/index.tsx]: done

**Comment by PaulFarault** [comment]
@gboutry 
- Operation_name:
  - Pressing spacebar in the middle will confirm   
    **This is done on purpose, as we don't want user to put spaces in an operation name.**
  - We cannot remove the generated input field when its empty   
    ✅ **Done**
  - Clicking outside leaves the field editable, therefore I'm not sure if my input is taken into account   
    ✅ **Done**
- Filter:
  - I believe we should be able to at least validate a regex   
    **Every input is considered as a string and therefore should appear as a valid regex.**
- Restart:
  - Restart should be set to False by default   
    ✅ **Done**

The validation toast could at least show the deployment id.   
IMO it could be improved after #54

**Comment by gboutry** [comment]
> @gboutry 
> - Operation_name:
>   - Pressing spacebar in the middle will confirm   
>     **This is done on purpose, as we don't want user to put spaces in an operation name.**
>   - We cannot remove the generated input field when its empty   
>     ✅ **Done**
>   - Clicking outside leaves the field editable, therefore I'm not sure if my input is taken into account   
>     ✅ **Done**
> - Filter:
>   - I believe we should be able to at least validate a regex   
>     **Every input is considered as a string and therefore should appear as a valid regex.**
> - Restart:
>   - Restart should be set to False by default   
>     ✅ **Done**
> 
> The validation toast could at least show the deployment id.   
> IMO it could be improved after #54 

What do you mean by "every input is considered a valid string", what if a user sends `hdfs_(\w+\)`, this is not a valid regex

**Comment by PaulFarault** [comment]
> What do you mean by "every input is considered a valid string", what if a user sends hdfs_(\w+\), this is not a valid regex

@gboutry done!

**Review by jpsavy — APPROVED** [review]
Test deploy from sources with the zookeeper_config : OK
Test filter : 
regex : OK
glob : OK

**Review by gboutry — APPROVED** [review]
LGTM

**Pull request merged**